process_report/invoices/bu_internal_invoice.py

```diff
@@ -0,0 +1,68 @@
+from dataclasses import dataclass
+from decimal import Decimal
+
+import process_report.invoices.invoice as invoice
+
+
+@dataclass
+class BUInternalInvoice(invoice.Invoice):
+    subsidy_amount: int
+
+    def _prepare(self):
+        def get_project(row):
+            project_alloc = row[invoice.PROJECT_FIELD]
+            if project_alloc.rfind("-") == -1:
+                return project_alloc
+            else:
+                return project_alloc[: project_alloc.rfind("-")]
+
+        self.data = self.data[
+            self.data[invoice.INSTITUTION_FIELD] == "Boston University"
+        ].copy()
+        self.data["Project"] = self.data.apply(get_project, axis=1)
+        self.data[invoice.SUBSIDY_FIELD] = Decimal(0)
+        self.data = self.data[
+            [
+                invoice.INVOICE_DATE_FIELD,
+                invoice.PI_FIELD,
+                "Project",
+                invoice.COST_FIELD,
+                invoice.CREDIT_FIELD,
+                invoice.SUBSIDY_FIELD,
+                invoice.BALANCE_FIELD,
+            ]
+        ]
+
+    def _process(self):
+        project_list = self.data["Project"].unique()
+        data_no_dup = self.data.drop_duplicates("Project", inplace=False)
+        sum_fields = [invoice.COST_FIELD, invoice.CREDIT_FIELD, invoice.BALANCE_FIELD]
+        for project in project_list:
+            project_mask = self.data["Project"] == project
+            no_dup_project_mask = data_no_dup["Project"] == project
+
+            sum_fields_sums = self.data[project_mask][sum_fields].sum().values
+            data_no_dup.loc[no_dup_project_mask, sum_fields] = sum_fields_sums
+
+        self.data = self._apply_subsidy(data_no_dup, self.subsidy_amount)
+
+    def _apply_subsidy(self, dataframe, subsidy_amount):
```

Comment:
I think this is a great opportunity to try and reuse the logic from applying credits here.

Reply:
I have added a new function in `util.py` that contains the shared logic between the `apply_subsidy` and `apply_new_pi_credit` functions.

Reply:
@QuanMPhm I think a better location for that function would be a common parent class for both invoice classes.

Reply:
@knikolla Could I put the function in the base `Invoice`, since I think the Billable and BU Internal invoice have much "relatedness" to warrant making a new parent class just to implement them. I think applying a flat discount would be something any invoice might want to do

Reply:
I would prefer not. While other invoices might want to implement a flat discount, the subsidies and credits invoices ALWAYS do.

Reply:
@QuanMPhm what are next steps here?

Reply:
I'm splitting up the commit for this PR to better show which commit copies over features, and which does more detailed refactoring.